---
title: "cgMLST R Workflow"
author: "Jimmy Liu and Eduardo Taboada"
date: "2023-05-03"
output:
  html_document:
    keep_md: yes
editor_options: 
  chunk_output_type: inline
---

# Background

The primary objective of this tutorial will be to provide a practical overview of core genome multilocus sequencing typing (cgMLST), a genome-based typing method widely adopted in bacterial genomic surveillance networks across the globe. The widespread adoption of cgMLST is due in large part to the recent advances in high-throughput sequencing, which have enabled the routine adoption of whole genome sequencing (WGS) in microbial surveillance. With WGS becoming increasingly accessible, part of the cascading effect was the development of efficient computational algorithms that further expedited the establishment of multilocus sequence typing at genome scales as routine practice.

In this exercise, you will use a custom workflow written in the R language to establish high quality cgMLST profiles that will be analyzed to infer genetic relatedness and explore clustering patterns in the genomic and contextual data.

To start, you are provided with pre-computed cgMLST allele calls generated by chewBBACA ([Silva et al. 2018](https://pubmed.ncbi.nlm.nih.gov/29543149/)), using a *E. coli* core genome scheme comprised of 2,360 loci downloaded from [here](https://zenodo.org/record/1323690). The following code template was used to call `chewBBACA` on a set of bacterial genomes consolidated in a single directory called `/path/to/my_genomes`:

``` bash
chewBBACA.py AlleleCall -i /path/to/my_genomes/ -g /path/to/cgMLST_scheme/ -o /path/to/results
```

## Learning Objectives

-   Understanding the significance of the metrics used for MLST-based quality control
-   Explaining the difference between core, accessory and pan-genome loci and their use in MLST analysis
-   Building dendrograms from MLST data
-   Familiarisation with the ggtree R library for tree visualization
-   Exploring the clonality of bacterial populations
-   Compare topological differences between SNV and cgMLST trees

## Analysis Dataset

The *E. coli* dataset has been uploaded to Google Sheets:

1.  [Metadata](https://docs.google.com/spreadsheets/d/1Mfqn07Zi9Dp0YgclSR4-Ohic7Fe3g2_dBlVjelZeDVI/edit?usp=sharing#gid=0)
2.  [cgMLST data](https://docs.google.com/spreadsheets/d/1Mfqn07Zi9Dp0YgclSR4-Ohic7Fe3g2_dBlVjelZeDVI/edit?usp=sharing#gid=962999752)

# Technical Notes

Many aspects of this R workflow have been prebaked in the sense that the code relies on specific input data formats and file paths. Thus, it can be rather difficult to rerun the code on a different dataset or in a different compute environment. It would be a shame if the utility of the R code presented here cannot be extended beyond the context of this tutorial, and hence we have prepared a technical guide below on how to reuse the Rmarkdown outside of the prebaked tutorial.

#### cgMLST Profiles

-   MLST calls must be stored in a **tab delimited** file called `cgMLST.tsv` under the `data/` folder.
-   The first column in the MLST calls must contain the genome IDs. The column header for the genome IDs must be named `#Name`.
-   The genome IDs in the `#Name` column should match the IDs in the metadata file. However, the order of the IDs in which they are recorded in the MLST profiles do not have to be identical to that of the metadata.
-   Unassigned/missing alleles must be encoded as `0` in MLST profiles

#### Metadata

-   Metadata must be stored in a **tab delimited** file called `metadata.tsv` under the `data/` folder.
-   The first column in the metadata must contain the genome IDs. The column header for the genome IDs must be named `ID`.
-   The metadata has to contain a `serovar` column, otherwise parts of the code that analyzes serovar information will fail. Serovar information may not be relevant to all organisms in which case, one can include null or NA values for the `serovar` column in the metadata.

#### Helper R scripts

-   The R code requires three helper scripts called `ggtree_helper.R`, `mlst_helper.R`, `cluster_helper.R`. All three scripts must be placed under the `src/` folder.

#### Genome annotations

-   To plot the presence and absence of genomic features (e.g. ARGs, VFs, Plasmids), you have to provide binary matrices for each feature type.
-   Binary matrices are similar to cgMLST profiles with genomes as rows and features as columns. Likewise, the first column in the binary matrices must contain the genome IDs and they must be formated as **tab delimited** files. The column header for the genome IDs must be named `ID`. The feature columns in the binary matrices must only contain `0` or `1` to indicate absence and presence, respectively.
-   These feature matrices must be placed under the `data/annotations/` folder.

#### Expected File Directory Structure

```         
/path/to/my_project/
  ├── cgMLST_workflow.Rmd
  ├── data/
  |   ├── cgMLST.tsv
  |   ├── metadata.tsv
  │   └── annotations/
  │       ├── Plasmids.tsv
  │       └── AMR.tsv 
  └── src/
      ├── ggtree_helper.R
      ├── cluster_helper.R
      └── mlst_helper.R
```

#### R Installation

-   The code should be compatible with R v4.0+
-   The recommended RStudio version is v1.4+
-   R library installation is automated when you run the first code chunk

# Getting Started

Let's begin by loading all R packages and helper scripts required to run all the code in this tutorial.

> [Every time you begin a new R session, you must reload all the packages and scripts!]{.underline}

```{r load pkgs and helper scripts, message = F, warning = F}
# install packages if unavailable
cran_pkgs <- c("tidyverse", "data.table", "ggnewscale", "BiocManager",
               "circlize", "phangorn", "randomcoloR", "RColorBrewer",
               "phytools", "knitr")
bioc_pkgs <- c("remotes", "ComplexHeatmap", "YuLab-SMU/treedataverse")
# check CRAN pkgs
for (i in cran_pkgs) { 
  if ( length(find.package(i, quiet = T)) == 0 ) {
    install.packages(i)
  }
}
# check Bioconductor pkgs
for (i in bioc_pkgs) { 
  if ( length(find.package(gsub(".*/", "", i), quiet = T)) == 0 ) {
    BiocManager::install(i)
  }
}
# load pkgs
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(treedataverse))
suppressPackageStartupMessages(library(ggnewscale))
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(randomcoloR))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(phangorn))
suppressPackageStartupMessages(library(phytools))
# load helper R scripts
source("src/mlst_helper.R")
source("src/ggtree_helper.R")
source("src/cluster_helper.R")
# set global chunk options
knitr::opts_chunk$set(echo = TRUE, warning = F)
```

Next, read the cgMLST data and metadata into memory using `fread()` from the `data.table` package.

```{r load data}
# data paths
metadata_path <- "data/metadata.tsv"
cgmlst_path <- "data/cgMLST.tsv"
# read file
metadata <- fread(metadata_path, sep = "\t", colClasses = "character")
cgmlst <- fread(cgmlst_path, sep = "\t")
```

# cgMLST QC

Before proceeding any further, you will first evaluate cgMLST data quality. A number of quality criteria will be defined to determine whether the properties of the allele profiles meet quality standards. Incomplete genome assembly and reference bias during MLST scheme construction can introduce significant levels of missing information in MLST data in the form of unassigned alleles at various loci. Hence, care must be taken to avoid comparisons that include loci and genomes with excessive numbers of unassigned alleles that could reduce the precision of genetic similarity calculations to be used for generating dendrograms for inferring relationships between the various genomes in the dataset.

Below, you will conduct a series of steps to compute the frequency of unassigned alleles across each locus (i.e. columns) and genome (i.e. rows), which will inform the identification of poor quality loci/genomes that may need to be flagged for removal from downstream analyses.

<img src="https://raw.githubusercontent.com/bioinformaticsdotca/IDE_2023/main/module5/img/mlst_qc_flowchart.jpg" width="750"/>

> [Editable parts of the code have been highlighted in the code chunks.]{.underline} You are highly encouraged to adjust them to observe how different parameters affect the outcome.

## Locus Quality

```{r compute loci completeness}
# use compute_lc helper function to compute 
# allele assignment rate (completeness) 
# across all loci
loci_completeness <- compute_lc(cgmlst)
# write to file
write.table(loci_completeness, "cgmlst_loci_quality.stats.tsv",
            quote = F, row.names = F, sep = "\t")
# print summary statistics for locus completeness
summary(loci_completeness)
```

```{r low assignment loci}
##### EDITABLE VARIABLE #####
lqual_threshold <- 2
#############################

# identify low qual loci
lqual_loci <- loci_completeness %>% 
  filter(missing_alleles > lqual_threshold) %>% 
  pull(locus)
# remove low qual loci from input cgMLST data
cgmlst_lc <- cgmlst %>% select(-all_of(lqual_loci))
# write to file
write.table(cgmlst_lc, "cgmlst_lqual_loci_rm.tsv",
            quote = F, row.names = F, sep = "\t")
# print filtering results
message(paste("Number of loci before filter:", ncol(cgmlst)-1))
message(paste("Number of loci after filter:", ncol(cgmlst_lc)-1))
message(paste("Number of loci removed:", ncol(cgmlst)-ncol(cgmlst_lc)))
message(paste0("Loci with ", lqual_threshold, " or less missing alleles were retained"))
```

## Defining Core Genes

```{r identify core loci}
##### EDITABLE VARIABLE #####
core_threshold <- 1
genome_qual <- 23
#############################

# compute core loci
core_loci <- calculate_core(
  mlst = cgmlst_lc,
  core_threshold = core_threshold,
  genome_qual = genome_qual
)
```

## Genome Quality

```{r genome completeness}
# compute genome completeness
# given quality filtered scheme
genome_completeness <- cgmlst_lc %>% 
  select(1, all_of(core_loci)) %>% 
  compute_gc()
# write to file
write.table(genome_completeness, "cgmlst_genome_qual.stats.tsv",
            quote = F, row.names = F, sep = "\t")
# print data summary of genome completeness
summary(genome_completeness)
```

```{r low assignment genomes}
##### EDITABLE VARIABLE #####
lqual_g_threshold <- 23
#############################

# identify low qual genomes
lqual_genomes <- genome_completeness %>% 
  filter(missing_alleles > lqual_g_threshold) %>% 
  pull(ID)
# remove low qual genomes from quality filtered cgmlst
cgmlst_final <- cgmlst_lc %>% 
  filter(!(`#Name` %in% lqual_genomes)) %>% 
  select(1, all_of(core_loci))
# remove low qual genomes from metadata
metadata <- metadata %>% filter(!(ID %in% lqual_genomes))
# write to file
write.table(cgmlst_final, "cgmlst_final.tsv",
            quote = F, row.names = F, sep = "\t")
# print filtering results
message(paste("Number of genomes before filter:", nrow(cgmlst_lc)))
message(paste("Number of genomes after filter:", nrow(cgmlst_final)))
message(paste("Number of genomes removed:", nrow(cgmlst_lc)-nrow(cgmlst_final)))
message(paste0("Genomes with ", lqual_g_threshold, " or less missing alleles were retained"))
```

# Hamming Distance

Distance-based and character-based methods can both be used to construct dendrograms from cgMLST data. However, the scope of this lab will only cover distance-based dendrograms, as character-based methods are covered extensively in other modules. In phylogenetic analysis, distance-based approaches are rather flexible in the sense that they can be constructed from any measure that quantifies genetic similarity, including distances computed by alignment-free (e.g. Mash) or alignment-based (e.g. BLAST) similarity search algorithms. Below, you are introduced to a metric called "Hamming distance", which is based on computing the number of differences between a pair of character vectors.

Given two character vectors of equal lengths, hamming distance is the total number of positions in which the two vectors are [different:]{.underline}

Profile A: `[ 0 , 2 , 0 , 5 , 5 , 0 , 0 , 0 , 0 ]`

Profile B: `[ 0 , 1 , 0 , 4 , 3 , 0 , 0 , 0 , 0 ]`

A != B:   `[ 0 , 1 , 0 , 1 , 1 , 0 , 0 , 0 , 0 ]`

Hamming distance = `sum( A != B )` = 3

In the context of two cgMLST profiles, hamming distance can be calculated based on the number of allele differences across all loci. Hamming distances will be computed in an all vs all fashion to generate a pairwise distance matrix that will serve as the input for distance-based tree-building algorithms, such as UPGMA and Neighbour-joining.

```{r calculate hamming distance}
dist_mat <- cgmlst_final %>% 
  column_to_rownames("#Name") %>% 
  t() %>% 
  hamming()
# print matrix dimension
# the dimension should be symmetric!
dim(dist_mat)
```

Let's now visualize the clustering patterns in the distance matrix in the form of a heatmap using the `ComplexHeatmap` package. We will also overlay serovar information to examine inter- and intra-serovar distances.

```{r distance matrix viz}
# create column annotations for heatmap
# to display serovar information
heatmap_annot <- metadata$serovar
names(heatmap_annot) <- metadata$ID
heatmap_annot <- heatmap_annot[order(factor(names(heatmap_annot),
                                            levels = rownames(dist_mat)))]
# create heatmap
dist_mat %>% 
  Heatmap(
    name = "cgMLST\nDistance",
    show_row_names = F, # do not display row labels
    show_column_names = F, # do not display column labels
    # use custom color gradient
    col = colorRamp2(
      c(min(dist_mat), mean(dist_mat), max(dist_mat)),
      c("#7ece97", "#eebd32", "#f76c6a")
    ),
    # add column annotation to show serovar info
    top_annotation = HeatmapAnnotation(
      Serovar = heatmap_annot,
      col = list(
					"Serovar" = structure(brewer.pal(length(unique(heatmap_annot)), "Set1"),
																names = unique(heatmap_annot))
      )
    )
  )
```

# Dendrogram Construction

Here you will construct a neighbour-joining (NJ) tree using the `nj()` function from the ape package. Alternatively, a UPGMA tree can be constructed by simply replacing `method = 'nj'` with `method = 'upgma'` in the code chunk below.

To visualize the resulting dendrogram, you will interact with the R package `ggtree`, which offers an extensive suite of functions to manipulate, visualize, and annotate tree-like data structures. In this section, you will be introduced to the different visual capabilities of the `ggtree` package, and progressively update the same tree with several layers of visual annotations based on available metadata.

## A Simple Tree Vis

```{r build distance tree, fig.height = 6}
# construct a core genome tree 
# using nj algorithm
cg_tree <- distance_tree(
  matrix = dist_mat,
  method = "nj"
)
# plot unrooted tree
cg_tree %>% 
  ggtree() +
  geom_nodelab(aes(label = node)) +
  geom_tiplab() +
  xlim(NA, 2200)
```

There are numerous ways in which one can root a tree. Here, we will root the cgMLST tree at a specific internal node to be consistent with the SNV tree.

```{r root tree, fig.height = 6}
# midpoint root
rooted_tree <- phangorn::midpoint(cg_tree)

# root by outgroup
outgroup <- "ED603"
outgroup_id <- as_tibble(cg_tree)$node[which(as_tibble(cg_tree)$label == outgroup)]
rooted_tree <- phytools::reroot(cg_tree, node.number = outgroup_id)

# root at an internal node
rooted_tree <- phytools::reroot(cg_tree, node.number = 76)

# export tree in newick format
write.tree(rooted_tree, file = "cgMLST_tree.nwk")

# plot rooted tree
rooted_tree %>% 
  ggtree() +
  geom_tiplab() +
  geom_rootpoint() +
  xlim(NA, 2500)
```

To start, run the following code chunk to plot a circular tree of the entire dataset with the tree tips colored by serovar information.

> You can assign a different metadata field to the `color_var` variable to update the mapping of the color aesthetics in the tree. For example setting `color_var = "serovar"` will color the tree tips by the serovar assignments

```{r draw cgMLST tree}
##### EDITABLE VARIABLE #####
color_var <- "serovar"
#############################

# set random seed
set.seed(123)
# determine category count
# in the color aes variable
n_colors <- length(unique(pull(metadata, !!sym(color_var))))

# plot core genome tree and
# colouring the tree tips 
# by color_var
rooted_tree %>% 
  ggtree() %<+% metadata +
  geom_tippoint(aes(color = !!sym(color_var)),
                size = 2) +
  guides(color = guide_legend(override.aes = list(size = 3) ) ) +
  scale_color_manual(values = distinctColorPalette(n_colors))
```

## Clustering by Distance

Identifying clusters of genomes sharing highly similar cgMLST profiles through the application of distance thresholds is a common practice in genomic surveillance and epidemiological investigations. Detecting novel clusters, comprising pathogen isolates from human clinical cases, can signal the emergence of an outbreak requiring a public health response and can provide important epidemiological insights on outbreak progression. Similarly, the co-clustering of outbreak isolates with isolates from food/environmental sources can help link the outbreak to possible sources/reservoirs of the pathogen in order to inform prevention and control measures.

In this section, you will generate genomic clusters from the dataset by applying several distance cutoffs. You will then annotate the dendrogram with cluster assignments to analyze cluster membership dynamics in response to distance cutoffs.

```{r cluster by distance, results = 'hide'}
# define clustering distance cutoffs
dist_cutoff <- seq(from=100, to=2000, by=100)
# perform complete linkage clustering
hclust_res <- purrr::map(dist_cutoff, function(x) {
  dist_mat %>% 
    as.dist() %>% 
    hclust(method = "complete") %>% 
    cutree(h = x) %>%
    as.factor()
})
names(hclust_res) <- paste0("clust_", dist_cutoff)

# print clustering results table
( 
  clusters <- data.frame(hclust_res) %>% 
    rownames_to_column("ID") 
)
```

## Cluster Analysis

Let's now add the clustering information to the previous tree and examine whether the above code chunk has generated sensible cluster assignments. Use the `plot_tree` function to analyze the clustering patterns in the core genome tree at different distance cutoffs.

```{r add cluster info, fig.height = 6, message = F}
plot_tree(
  tree = rooted_tree,
  clusters = clusters,
  color_by = "serovar", # which metadata variable to color tree tips by?
  color.tiplab = F, # whether to color tip labels
  tip.size = 3, # size of tree tip point
  legend.x = 0.15, # legend position on x axis
  legend.y = 0.8, # legend position on y axis
  legend.size = 5, # legend text size
  legend.ncat.per.col = 8, # number of categories to show per column in legend
  legend.hide = F, # whether to hide colour legend
  plot.xlim = 2800, # plot area width
  label_vars = c("ID"), # metadata vars only
  label.offset = 25, # distance between labels and tree tips
  label.size = 4, # tree tip label text size
  label.align = F, # align tip labels vertically
  annot.offset = 0.12, # distance between heatmap and tree tips
  annot.width = 0.15, # heatmap width
  annot.textsize = 4, # heatmap x axis text label size
  annot.ncol = 6 # number of annotation columns to display
)
# export tree in pdf
ggsave("cgMLST_tree.pdf", height = 8, width = 8)
```

## Comparing SNV and cgMLST Analysis

Here, we will compare the topologies of the cgMLST distance tree to the SNV phylogeny produced from the earlier module. We will use the `read.tree()` function from the `ggtree` package to load the SNV tree in newick format into the R environment, and use `plot_tree()` function as above to visualize the tree.

Both trees are saved as PDF format. Using any PDF viewer, open up the two files side by side, and look for similarities and differences in tree topology and branch lengths.

```{r SNV tree, fig.height = 7, message = F, warning = F}
snv_tree_path <- "data/vtec_snv_tree.nwk"

# read tree
snv_tree <- read.tree(snv_tree_path)

# plot snv tree
plot_tree(
  tree = snv_tree,
  clusters = clusters,
  color_by = "serovar", # which metadata variable to color tree tips by?
  color.tiplab = F, # whether to color tip labels
  tip.size = 3, # size of tree tip point
  legend.x = 0.1, # legend position on x axis
  legend.y = 0.85, # legend position on y axis
  legend.size = 5, # legend text size
  legend.ncat.per.col = 7, # number of categories to show per column in legend
  legend.hide = F, # whether to hide colour legend
  plot.xlim = 1.4, # plot area width
  label_vars = c("ID"), # metadata vars only
  label.offset = 0.02, # distance between labels and tree tips
  label.size = 4, # tree tip label text size
  label.align = F, # align tip labels vertically
  annot.offset = 0.13, # distance between heatmap and tree tips
  annot.width = 0.15, # heatmap width
  annot.textsize = 4, # heatmap x axis text label size
  annot.ncol = 6 # number of annotation columns to display
)
# export tree in pdf
ggsave("SNV_tree.pdf", height = 8, width = 8)
```

## Integrating Genome Annotations

The `plot_tree()` function is a very powerful function that goes beyond analyzing genomic clusters and contextual data. It is also able to integrate genome annotation information to tree visualizations to identify genomic features such as VFs or ARGs that are localized to specific clusters. The genome annotation information can be provided to `plot_tree()` using the `annotations` parameter. You can also choose which feature type(s) to display using the `annotations_var` parameter.

The annotation information must be formatted as a data frame that consists of four columns: \* `ID`: genome IDs, \* `feature`: name of the genomic feature such as gene name \* `value`: 0 or 1 indicating absence or presence of feature \* `type`: the type of genomic feature such as AMR, Plasmids, VFs

The code chunk below will search for the presence/absence matrices (tab delimited files) stored under `data/annotations/` and parse the file content to generate a genome annotation table in the format described above.

```{r load annotations}
annotation_dir <- "data/annotations/"
annotation_types <- str_replace(list.files(annotation_dir, ".tsv"), ".tsv", "")

annotations <- map_dfr(annotation_types, function(x) {
  
  df <- fread(file.path(annotation_dir, paste0(x, ".tsv")),
              colClasses = "character")
  # identify columns with only 1 unique value
  cols <- which(unlist(lapply(df, function(x) return(length(unique(x)) != 1))))
  # remove cols with only 1 unique value
  df %>% 
    select(cols) %>% 
    pivot_longer(cols = 2:ncol(.),
                 names_to = "feature",
                 values_to = "value") %>%
    mutate(type = x,
           value = as.character(value))
    
})

annotations
```

```{r integrated tree, fig.height=10, message = F, warning = F}
plot_tree(
  tree = rooted_tree,
  clusters = clusters,
  clusters.show = T, # whether to show clustering data
  annotations = annotations,
  annotations_var = c("Drug Classes", "VFs", "Plasmids"),
  color_by = "serovar", # which metadata variable to color tree tips by?
  color.tiplab = F, # whether to color tip labels
  tip.size = 5, # size of tree tip point
  legend.x = 0.08, # legend position on x axis
  legend.y = 0.75, # legend position on y axis
  legend.size = 7, # legend text size
  legend.ncat.per.col = 10, # number of categories to show per column in legend
  legend.hide = F, # whether to hide colour legend
  plot.xlim = 7000, # plot area width
  label_vars = c("ID"), # metadata vars only
  label.offset = 120, # distance between labels and tree tips
  label.size = 6, # tree tip label text size
  label.align = T, # align tip labels vertically
  annot.textsize = 6, # heatmap x axis text label size
  annot.offset = c(0.3, 0.07, 0.07, 0.07), # distance between heatmap and tree tips
  annot.width = c(0.35, 0.1, 0.5, 0.5), # heatmap width
  annot.ncol = c(6, 2, 10, 10), # number of annotation columns to display
  annot.random.sample = F, # randomly select annotation columns to display
  annot.random.seed = 3269, # specify a random seed for reproducibility
  bottom.margin = 2.5 # width of the bottom margin
)

# save plot as pdf
ggsave("integrated_tree.pdf", height = 10, width = 20)
```
